### ISSUE TITLE
Deadlock: Navigation stops when request interception is enabled

### ISSUE THREAD

**Issue by inancgumus** [grafana/xk6-browser#1072]
`NetworkManager` deadlocks while racing with two different browser network events:
* `onLoadingFailed`
* `onRequestPaused`

This issue only happens when we have outgoing requests that are in [the k6-blocked hosts and IPs list](https://github.com/grafana/xk6-browser/blob/5c56565e1231f2d0d9423b46690ff7d5db352146/common/network_manager.go#L433-L523). The outgoing requests won't let other communications with the browser go through. When we remove the [`network.GetResponseBody`](https://github.com/grafana/xk6-browser/blob/5c56565e1231f2d0d9423b46690ff7d5db352146/common/response.go#L128-L132) call (called by [`emitResponseMetrics`](https://github.com/grafana/xk6-browser/blob/5c56565e1231f2d0d9423b46690ff7d5db352146/common/network_manager.go#L170-L251) when a request loading [finishes](https://github.com/grafana/xk6-browser/blob/5c56565e1231f2d0d9423b46690ff7d5db352146/common/network_manager.go#L372-L376) or we're handling a [redirect](https://github.com/grafana/xk6-browser/blob/5c56565e1231f2d0d9423b46690ff7d5db352146/common/network_manager.go#L253-L261)), the test run finishes.

We discovered how and why this problem occurs, ~~but the underlying reason why this is happening has yet to be entirely apparent~~. Find a way to reproduce this behavior with a Go test, discover the underlying reason, report it, and fix the problem.

**Update: Root cause discovered**

The main page redirects to another page before finishing its loading. The redirected page gets blocked by the network interception (_blocked hosts and IPs_) analysis stage. The problem was that we needed to tell the redirected page to continue. But we couldn't because we were waiting for the main page to load. Hence the deadlock.

### xk6-browser version

v1.1.0

### Steps to reproduce the problem

* Copy the script from the test run ID: 161224.
* Run the test script locally with a blocklist (no matter which).

### Relevant links

* [Bug: Request interception and caching may trigger double pause on stylesheet-initiated font requests in some conditions](https://bugs.chromium.org/p/chromium/issues/detail?id=1196004)
* [Search: Request Interception](https://bugs.chromium.org/p/chromium/issues/list?q=Request%20interception&can=2)
* [Search: requestPaused](https://bugs.chromium.org/p/chromium/issues/list?q=requestPaused&can=2)
* [Search: loadingFinished](https://bugs.chromium.org/p/chromium/issues/list?q=loadingFinished&can=2)

```[tasklist]
### Tasks
- [x] Find the root cause of the issue
- [x] Reproduce the issue with a Go test
- [x] https://github.com/grafana/xk6-browser/pull/1084
- [x] https://github.com/grafana/xk6-browser/pull/1078
- [x] https://github.com/grafana/k6/pull/3425
- [x] https://github.com/grafana/xk6-browser/pull/1086
- [x] Test the fix on remote
- [x] Make a k6 release on remote
- [x] Add an improvement issue (https://github.com/grafana/xk6-browser/issues/1093)
- [ ] https://github.com/grafana/k6/pull/3447
```

**Comment by inancgumus** [comment]
We receive and process the following and then send a `getResponseBody`:

```
<- {"method":"Network.loadingFinished","params":{"requestId":"51317.37","timestamp":24102.926893,"encodedDataLength":6759,"shouldReportCorbBlocking":false},"sessionId":"2D176DA231707745A2E7710694B2A890"}
-> {"id":147,"method":"Network.getResponseBody","params":{"requestId":"51317.37"}}
```

`getResponseBody` is intercepted by the network request blocker and waits for us to tell it to continue, but we never do because we're blocked in the same `NetworkManager` goroutine, calling `getResponseBody` in `loadingFinished`. The following is never processed:

```
<- {"method":"Network.requestWillBeSent","params": ...}
<- {"method":"Fetch.requestPaused","params":{"requestId":"interception-job-20.0" ...}
<- {"method":"Network.loadingFailed","params":{"requestId":"51317.38", ...}
```

**Comment by inancgumus** [comment]
https://chromedevtools.github.io/devtools-protocol/tot/Fetch/#method-getResponseBody

> Causes the body of the response to be received from the server and returned as a single string. **May only be issued for a request that is paused in the Response stage** and is mutually exclusive with takeResponseBodyForInterceptionAsStream. **Calling other methods that affect the request** or disabling fetch domain before body is received results in **an undefined behavior**.
>
> Note that the response body is not available for redirects. Requests paused in the redirect received state may be differentiated by responseCode and presence of location response header, see comments to requestPaused for details.

**Comment by inancgumus** [comment]
More exploration reveals that we might not be handling redirected request interception.

**Comment by inancgumus** [comment]
We've solved this issue in #1078. However, the solution can be suboptimal since it uses Go routines for each metric emission. This can be confusing when something goes wrong, and we want to analyze the logs (since goroutines dump their state on panic).

We can still find a better solution that carefully tracks network events. Then we can postpone fetching the body and do that when the correct time comes. I'll keep this issue open until we find a better solution.

cc: @grafana/k6-browser
